### PR TITLE
Graph freezing and unfreezing

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
This PR introduces two graph "actions":
- [x] A method to freeze the weights of (all/selected) trainable modules in a graph
- [x] A method to unfreeze the weights of (all/selected) trainable modules in a graph

In order to make it work “in an elegant” way they PR introduces a “back-end independent module type” (ModuleType enum).

### CODE REVIEW

File: nemo/core/neural_modules.py [hunk 1]
```diff
@@ -39,6 +38,15 @@
 YAML = YAML(typ='safe')
 
 
+class ModuleType(Enum):
+    """ Back-end independent module types """
+
+    module = 0
```
Comment (okuchaiev) [inline nemo/core/neural_modules.py:44]: can you please this set to: "trainable, nontrainable, datalayer, loss" ?

File: nemo/backends/pytorch/nm.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import torch as t
 import torch.nn as nn
 
-from nemo.core import DeviceType, NeuralModule, WeightShareTransform
+from nemo.core import DeviceType, ModuleType, NeuralModule, WeightShareTransform
```
Comment (okuchaiev) [inline nemo/backends/pytorch/nm.py:9]: If you introduce this enum, you also need to set this property in NonTrainableNM

File: nemo/core/neural_modules.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply nemo/core/neural_modules.py:44]: Good catch... but once again: I did that for a reason. I truly believe that we should get rid of a NonTrainableNM class. The only thing it is doing now it is stopping the gradient backpropagation - and **this is simply wrong**. 

Anyway, this is a subject for a different conversation. Added "nontrainable" enum.

File: nemo/backends/pytorch/nm.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply nemo/backends/pytorch/nm.py:9]: done
